dms_version/models/dms_version.py

```diff
@@ -0,0 +1,160 @@
+
+    name = fields.Char(string="Name", readonly=True)
+    file_id = fields.Many2one(
+        comodel_name="dms.file", ondelete="cascade", string="File", required=True
+    )
+    storage_id = fields.Many2one(
+        related="file_id.storage_id",
+        comodel_name="dms.storage",
+        string="Storage",
+        readonly=True,
+    )
+    compress_versions = fields.Boolean(
+        related="storage_id.compress_versions", readonly=True
+    )
+    incremental_versions = fields.Boolean(
+        related="storage_id.incremental_versions", readonly=True
+    )
+    previous_version = fields.Many2one(
+        comodel_name="dms.version", ondelete="set null", string="Previous Version"
+    )
+    next_version = fields.Many2one(
+        comodel_name="dms.version", ondelete="set null", string="Next Version"
+    )
+    is_compress = fields.Boolean(string="Compressed", default=False)
+    is_incremental = fields.Boolean(string="Incremental", default=False)
+    size = fields.Integer(string="Size", readonly=True)
+    content = fields.Binary(
+        compute="_compute_content",
+        inverse="_inverse_content",
+        string="Content",
+        attachment=False,
+        prefetch=False,
+        required=True,
+        store=False,
+    )
+    binary = fields.Binary(
+        string="Binary", prefetch=False, attachment=True, invisible=True
+    )
+
+    def get_binary(self, patch=True):
+        self.ensure_one()
+        binary = base64.b64decode(self.binary)
+        if self.is_compress:
+            binary = zlib.decompress(binary)
+        if self.is_incremental and patch:
+            prev_version = self.previous_version
+            prev_binary = prev_version.get_binary()
+            binary = bsdiff4.patch(prev_binary, binary)
+        return binary
+
+    @api.model
+    def clean_versions(self, dms_files, max_size=None):
+        for dms_file in dms_files.filtered(lambda rec: rec.count_versions > 0):
+            max_size = max_size or dms_file.storage_id.stored_versions
+            if dms_file.count_versions > max_size:
+                remove_size = dms_file.count_versions - max_size
+                remove_versions = self.env["dms.version"]
+                current = dms_file.version_ids.filtered(
+                    lambda rec: not rec.previous_version
+                )
+                while len(remove_versions) < remove_size:
+                    remove_versions |= current
+                    current = current.next_version
+                current.write(
+                    {
+                        "content": current.content,
+                        "previous_version": None,
+                        "is_incremental": False,
+                    }
+                )
+                remove_versions.unlink()
+
+    @api.model
+    def add_version(self, dms_file, content):
+        domain = [
+            ("file_id", "=", dms_file.id),
+            ("next_version", "=", False),
+        ]
```

Comment:
It doesn't seem to make sense to use an `@api.model` method for something that expects a record. Instead:

```suggestion
    def add_version(self, content):
        domain = [
            ("file_id", "=", self.id),
            ("next_version", "=", False),
        ]
```